### PR TITLE
Upgrade SWIG to support NodeJS v12 / V8 v7.6, v8 v7.8 runtime

### PR DESCRIPTION
**Pull request by lissyx**
WIP, trying to get Travis to run ?

### CODE REVIEW

**Comment by yegorich** [comment]
@lissyx I could successfully test your PR on ARM system with Node.js  version: v12.16.0.

**Comment by lissyx** [comment]
> @lissyx I could successfully test your PR on ARM system with Node.js version: v12.16.0.

Thanks for that, even though I knew it worked :). What I'm really needing right now is jsut some help to make the PR / travis-ci itself work.

**Comment by yegorich** [comment]
@lissyx I have no experience with v8. I'm just using SWIG for a small library. 

Your patch touches a lot of files. What about dividing it into several logical patches/PRs? One for fixes, one for tests, and one for tools enhancements with detailed commit logs? This would simplify the review process.

Perhaps @p2k and @furylynx could help?

**Comment by lissyx** [comment]
> @lissyx I have no experience with v8. I'm just using SWIG for a small library.

I don't either.



>  What about dividing it into several logical patches/PRs? One for fixes, one for tests, and one for tools enhancements with detailed commit logs? This would simplify the review process.

Problem is not reviewing it, problem is being able to run on travis to ensure it is working as intended. I have spend a few hours on that and only failed so far because of nodejs + travis specifics I'm not familiar with. Also, it took ~5h each time for my PR to start and a few more hours to complete, so this is definitively not something I can spend more time on.

If you are willing to take the PR and get it to run on travis / for review, please, don't hesitate.

**Comment by yegorich** [comment]
Do you have issues only with Travis-CI or also locally? On my Ubuntu 18.04.3 LTS with Node v12.16.1, I have issues when invoking:

    make check-javascript-examples

Most prominent ones are like this (`Examples/javascript/native`):

```
../example_wrap.cxx: In function ‘SwigV8ReturnValue JavaScript_do_work(const SwigV8Arguments&)’:
../example_wrap.cxx:1668:13: error: ‘Handle’ is not a member of ‘v8’
         v8::Handle<v8::Value> jsresult =
             ^~~~~~
../example_wrap.cxx:1668:29: error: expected primary-expression before ‘>’ token
         v8::Handle<v8::Value> jsresult =
                             ^
../example_wrap.cxx:1668:31: error: ‘jsresult’ was not declared in this scope
         v8::Handle<v8::Value> jsresult =
                               ^~~~~~~~

```

**Comment by lissyx** [comment]
> Do you have issues only with Travis-CI or also locally? On my Ubuntu 18.04.3 LTS with Node v12.16.1, I have issues when invoking:

It was working for me locally, but I have not let it run completely. I also had to pass a few more env variables to make (I think it was to ensure it uses the proper swig version) but I don't remember which ones.


> Most prominent ones are like this (`Examples/javascript/native`):

As stated elsewhere, this patch might not be complete. Getting it to run on Travis was the first requirement to be able to ensure that it could be verified to cover all use-cases. Fixing those if it's incomplete should not be very hard, though, it's likely the same reasoning.

**Comment by yegorich** [comment]
So with this change, I have got it running on my system:

```
diff --git a/Examples/javascript/native/example.i b/Examples/javascript/native/example.i
index 8c6160060..e94f18118 100644
--- a/Examples/javascript/native/example.i
+++ b/Examples/javascript/native/example.i
@@ -15,7 +15,7 @@ int placeholder() { return 0; }
     static SwigV8ReturnValue JavaScript_do_work(const SwigV8Arguments &args) {
         SWIGV8_HANDLESCOPE();
         const int MY_MAGIC_NUMBER = 5;
-        v8::Handle<v8::Value> jsresult =
+        v8::Local<v8::Value> jsresult =
             SWIG_From_int(static_cast< int >(MY_MAGIC_NUMBER));
         if (args.Length() != 0)
             SWIG_exception_fail(SWIG_ERROR, "Illegal number of arguments.");

```

**Comment by lissyx** [comment]
> So with this change, I have got it running on my system:
> 
> ```
> diff --git a/Examples/javascript/native/example.i b/Examples/javascript/native/example.i
> index 8c6160060..e94f18118 100644
> --- a/Examples/javascript/native/example.i
> +++ b/Examples/javascript/native/example.i
> @@ -15,7 +15,7 @@ int placeholder() { return 0; }
>      static SwigV8ReturnValue JavaScript_do_work(const SwigV8Arguments &args) {
>          SWIGV8_HANDLESCOPE();
>          const int MY_MAGIC_NUMBER = 5;
> -        v8::Handle<v8::Value> jsresult =
> +        v8::Local<v8::Value> jsresult =
>              SWIG_From_int(static_cast< int >(MY_MAGIC_NUMBER));
>          if (args.Length() != 0)
>              SWIG_exception_fail(SWIG_ERROR, "Illegal number of arguments.");
> ```

So it would be the example itself that is broken and not the bindings? Anyway, again, don't hesitate to steal my PR, I don't really have time those days to debug that :/

File: Examples/test-suite/javascript/abstract_access_runme.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
-var abstract_access = require("abstract_access");
+var abstract_access = require("./abstract_access");
 
 var d = new abstract_access.D()
 if (d.do_x() != 1) {
    throw "Error";
 }
+
+process.exit(0);
```
Comment (yegorich) [inline Examples/test-suite/javascript/abstract_access_runme.js]: What is the reason for explicit exit call? AFAIK if successful, the process always exits with 0.

File: Examples/test-suite/javascript/abstract_access_runme.js [hunk 2]
```diff
@@ -1,6 +1,8 @@
-var abstract_access = require("abstract_access");
+var abstract_access = require("./abstract_access");
```
Comment (yegorich) [inline Examples/test-suite/javascript/abstract_access_runme.js]: Why this change? Is it because of electron support?

File: Examples/test-suite/javascript/abstract_access_runme.js (hunk 1 above)
Comment (lissyx) [reply Examples/test-suite/javascript/abstract_access_runme.js]: That's my assumption as well, but on electron the process would not exit unless this call. I lack knowledge of electron to know if that's right behavior though, but that felt like an acceptable midway step until I could get a proper review done.

File: Examples/test-suite/javascript/abstract_access_runme.js (hunk 2 above)
Comment (lissyx) [reply Examples/test-suite/javascript/abstract_access_runme.js]: Yes, it seems in some context it would not find modules even though it is at the right place. Electron, as you guessed. Again, maybe there is a better solution but I figured it would be acceptable to at least start getting tests green on Travis.

Comment (lissyx) [reply Examples/test-suite/javascript/abstract_access_runme.js]: Yes, it seems in some context it would not find modules even though it is at the right place. Electron, as you guessed. Again, maybe there is a better solution but I figured it would be acceptable to at least start getting tests green on Travis.

**Comment by yegorich** [comment]
@lissyx  as far as I can see, you also haven't found a solution for

`error: ‘bool v8::Object::Set(v8::Local<v8::Value>, v8::Local<v8::Value>)’ is deprecated: Use maybe version [-Werror=deprecated-declarations]`

**Comment by lissyx** [comment]
> @lissyx as far as I can see, you also haven't found a solution for
> 
> `error: ‘bool v8::Object::Set(v8::Local<v8::Value>, v8::Local<v8::Value>)’ is deprecated: Use maybe version [-Werror=deprecated-declarations]`

Again, I've only covered what was required to get my module to work.

**Comment by yegorich** [comment]
I have found the solution. There was another `Set` invocation without `SWIGV8_CURRENT_CONTEXT()` as the first parameter in `Lib/javascript/v8/javascriptcode.swg`.

**Comment by lissyx** [comment]
> Aside from the problem with `SWIG_Object`, `make check-javascript-test-suite` runs without errors against Node.js v13.

Locally or on Travis ?

**Comment by yegorich** [comment]
@lissyx locally but I could try to amend my first PR and try it on Travis. But as long as we don't resolve the `SWIG_Object` issue, it doesn't really matter. Stay tuned.

**Comment by lissyx** [comment]
> @lissyx locally but I could try to amend my first PR and try it on Travis. But as long as we don't resolve the `SWIG_Object` issue, it doesn't really matter. Stay tuned.

Are you referring to that? https://github.com/lissyx/swig/commit/5e1bdcfd6e5bd8353993433bd0efb4a3cb3ca639#diff-49eeb9acfa050a3669197366d0e77627R28

**Comment by yegorich** [comment]
@lissyx Travis is satisfied. I've enabled both v12 and v13 for https://github.com/swig/swig/pull/1736. Removing Node.js older than v10 makes the code really simple and clean. There are only minor changes required to support v10 and we can switch to checking only `NODE_MODULE_VERSION`. See this [hint](https://nodejs.org/en/download/releases/#ref-1) for details.

**Comment by lissyx** [comment]
> @lissyx Travis is satisfied. I've enabled both v12 and v13 for #1736. Removing Node.js older than v10 makes the code really simple and clean. There are only minor changes required to support v10 and we can switch to checking only `NODE_MODULE_VERSION`. See this [hint](https://nodejs.org/en/download/releases/#ref-1) for details.

Yeah, I knew about that, but that was one of the things not really worth polishing until it was working well on Travis. Thanks for picking up this work and getting it further!

**Comment by vadz** [comment]
I think this one (and #1736?) is replaced by #1746 and can be closed, can't it?

**Comment by yegorich** [comment]
This PR is now obsolete. Let's close it.
